src/lib.rs

```diff
@@ -16,116 +16,261 @@
+        .filter_map(|res| {
+            let path = match res {
+                Ok(e) => e.path(),
+                Err(e) => panic!("failed to list {}: {:?}", out_dir, e),
+            };
+            if path.extension() == Some(std::ffi::OsStr::new("rs")) {
+                generated.push(format!("{}", path.display()));
+                let name = path.file_stem().unwrap().to_str().unwrap();
+                Some((name.replace('-', "_"), name.to_owned()))
+            } else {
+                None
+            }
+        })
+        .collect();
+    replace_read_unknown_fields(&generated);
+    let mut f = File::create(format!("{}/mod.rs", out_dir)).unwrap();
+    for (module, file_name) in &modules {
+        if !module.contains('.') {
+            writeln!(f, "pub mod {};", module).unwrap();
+            continue;
+        }
+        let mut level = 0;
+        for part in module.split('.') {
+            writeln!(f, "{:level$}pub mod {} {{", "", part, level = level).unwrap();
+            level += 1;
+        }
+        writeln!(f, "include!(\"{}.rs\");", file_name).unwrap();
+        for _ in (0..level).rev() {
+            writeln!(f, "{:1$}}}", "", level).unwrap();
+        }
     }
 }
 
-/// Use protobuf-rs to generate Rust files from proto files (`file_names`).
-///
-/// Uses `["proto", "include"]` as the include lists.
-pub fn generate_protobuf_files<T: AsRef<str>>(file_names: &[T], out_dir: &str) {
-    let file_names = &file_names.iter().map(AsRef::as_ref).collect::<Vec<_>>();
-    protoc_rust::run(protoc_rust::Args {
-        out_dir,
-        input: file_names,
-        includes: &["proto", "include"],
-        customize: protoc_rust::Customize {
-            ..Default::default()
-        },
-    })
-    .unwrap();
-
-    protoc_grpcio::compile_grpc_protos(file_names, &["proto", "include"], out_dir, None).unwrap();
+/// Use rust-protobuf to generate Rust files from proto files (`files`).
+#[cfg(feature = "protobuf-codec")]
+mod protobuf_imps {
+    use regex::Regex;
+    use std::env;
+    use std::fmt::Debug;
+    use std::path::Path;
+    use std::process::Command;
+    use std::str::from_utf8;
+
+    pub fn get_protoc() -> String {
+        let protoc_bin_name = match (env::consts::OS, env::consts::ARCH) {
+            ("linux", "x86") => "protoc-linux-x86_32",
+            ("linux", "x86_64") => "protoc-linux-x86_64",
+            ("linux", "aarch64") => "protoc-linux-aarch_64",
+            ("linux", "ppcle64") => "protoc-linux-ppcle_64",
+            ("macos", "x86_64") => "protoc-osx-x86_64",
+            ("windows", _) => "protoc-win32.exe",
+            _ => return "protoc".to_owned(),
+        };
+        let bin_path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
+            .join("bin")
+            .join(protoc_bin_name);
+        format!("{}", bin_path.display())
+    }
+
+    /// Check that the user's installed version of the protobuf compiler is 3.1.x.
+    pub fn check_protoc_version(protoc: &str) {
```

Comment:
I don't think we need this function any more - we will always use the bundled protoc

Reply:
It will fallback to protoc in PATH if the platform is not supported.